openeo-geotrellis/src/main/scala/org/openeo/geotrellis/package.scala

```diff
@@ -158,28 +160,67 @@ package object geotrellis {
     val seconds, days, undefined = Value
   }
 
-  def withRetries[R](action: => scalaj.http.HttpResponse[R]): scalaj.http.HttpResponse[R] = {
-    @tailrec
-    def attempt(retries: Int, delay: (Long, TimeUnit))(action: => scalaj.http.HttpResponse[R]): scalaj.http.HttpResponse[R] = {
-      val (amount, timeUnit) = delay
-
-      val ret = action
-      if (ret.code == 429) {
-        logger.warn(f"Got ${ret.code}. Will retry request: " + ret)
-        // TODO: Could use Retry-After header.
-        // It is not returned by https://services.terrascope.be/ at the moment.
-        // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After
-      } else {
-        return ret
-      }
-      timeUnit.sleep(amount)
-      attempt(retries - 1, (amount * 2, timeUnit)) {
-        action
-      }
+
+  /**
+   * taken from DefaultProcessApi
+   */
+  def withRetryAfterRetries[R](context: String)(fn: => HttpResponse[R])(implicit logger: Logger): HttpResponse[R] = {
+    var lastResponse: Option[HttpResponse[R]] = None
+
+    val retryable: Throwable => Boolean = {
+      case HttpStatusException(400, _, responseBody) if responseBody.contains("Request body should be non-empty.") => true
+      case HttpStatusException(statusCode, _, responseBody) if statusCode >= 500
+        && !responseBody.contains("newLimit > capacity") && !responseBody.contains("Illegal request to https") => true
+      case _: SocketTimeoutException => true
+      case _: CirceException => true
+      case e => logger.error(s"Not attempting to retry unrecoverable error in context: $context", e); false
     }
 
-    attempt(retries = 4, delay = (30, SECONDS)) {
-      action
+    val shakyConnectionRetryPolicy = new net.jodah.failsafe.RetryPolicy[HttpResponse[R]]()
+      .handleIf(retryable.asJava)
+      .withBackoff(1, 1000, ChronoUnit.SECONDS) // should not reach maxDelay because of maxAttempts 5
+      .withJitter(0.5)
+      .withMaxAttempts(5)
+      .onFailedAttempt((attempt: ExecutionAttemptedEvent[HttpResponse[R]]) => {
+        val e = attempt.getLastFailure
+        logger.warn(s"Attempt ${attempt.getAttemptCount} failed in context: $context", e)
+      })
+      .onFailure((execution: ExecutionCompletedEvent[HttpResponse[R]]) => {
+        val e = execution.getFailure
+        logger.error(s"Failed after ${execution.getAttemptCount} attempt(s) in context: $context", e)
+      })
+
+    val isRateLimitingResponse: Throwable => Boolean = {
+      case HttpStatusException(429, _, _) => true
+      case _ => false
     }
+
+    var retryAfterSecondsCounter = 20.0
```

Comment:
Unnecessary to keep state yourself, delay with back-off can be derived from `executionContext` argument of `withDelay` (see further on).